File: back/app/schemas/project.py
"""Schémas Pydantic pour les projets de rénovation (BAR-TH-175)."""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ProjectStatus(str, Enum):
    """Statut d'un projet de rénovation."""

    DRAFT = "DRAFT"
    IN_PROGRESS = "IN_PROGRESS"
    AUDIT_PENDING = "AUDIT_PENDING"
    VALIDATED = "VALIDATED"
    COMPLETED = "COMPLETED"
    ARCHIVED = "ARCHIVED"


# ============================================================================
# Schémas de base
# ============================================================================


class ProjectBase(BaseModel):
    """Champs communs pour un projet."""

    name: str = Field(..., min_length=1, max_length=255, description="Nom du projet")
    client_id: UUID | None = Field(None, description="ID du client (bailleur) associé")
    building_address: str | None = Field(
        None, max_length=500, description="Adresse de l'immeuble"
    )
    total_apartments: int | None = Field(
        None, ge=1, description="Nombre total d'appartements"
    )
    module_code: str = Field(
        default="BAR-TH-175",
        max_length=50,
        description="Code du module CEE",
    )
    data: Dict[str, Any] = Field(
        default_factory=dict, description="Données flexibles du projet"
    )


class ProjectCreate(ProjectBase):
    """Payload de création d'un projet."""

    pass


class ProjectUpdate(BaseModel):
    """Payload de mise à jour partielle d'un projet."""

    name: str | None = Field(None, min_length=1, max_length=255)
    client_id: UUID | None = None
    building_address: str | None = None
    total_apartments: int | None = Field(None, ge=1)
    status: ProjectStatus | None = None
    data: Dict[str, Any] | None = None


class ProjectResponse(ProjectBase):
    """Représentation API d'un projet."""

    id: UUID
    tenant_id: UUID
    status: ProjectStatus
    created_at: datetime
    updated_at: datetime
    archived_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class PaginatedProjectsResponse(BaseModel):
    """Réponse paginée pour la liste des projets."""

    items: List[ProjectResponse]
    total: int
    page: int
    page_size: int


# ============================================================================
# Schémas pour les opérations sur les projets
# ============================================================================


class BulkDraftsCreate(BaseModel):
    """Payload pour créer des appartements (ModuleDrafts) en masse."""

    quantity: int = Field(
        ...,
        ge=1,
        le=500,
        description="Nombre d'appartements à créer",
    )
    apartment_type: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="Type d'appartement (T1, T2, T3, T4, T5, Studio, etc.)",
    )
    common_data: Dict[str, Any] = Field(
        default_factory=dict,
        description="Données communes à appliquer à tous les appartements créés",
    )


class BulkDraftsResponse(BaseModel):
    """Réponse après création en masse."""

    created_count: int
    project_id: UUID
    draft_ids: List[UUID]


class PropagateAuditRequest(BaseModel):
    """Payload pour propager les données d'audit d'un appartement vers d'autres."""

    source_draft_id: UUID = Field(
        ..., description="ID du draft source contenant les données à propager"
    )
    target_draft_ids: List[UUID] = Field(
        ...,
        min_length=1,
        description="Liste des IDs de drafts cibles",
    )
    fields_to_propagate: List[str] | None = Field(
        None,
        description="Liste des champs à propager. Si None, propage toutes les données.",
    )


class PropagateAuditResponse(BaseModel):
    """Réponse après propagation d'audit."""

    updated_count: int
    updated_draft_ids: List[UUID]
    skipped_draft_ids: List[UUID] = Field(
        default_factory=list,
        description="IDs des drafts non mis à jour (introuvables ou hors projet)",
    )

File: front/src/types/project.ts
// ============================================================================
// Enums
// ============================================================================

export type ProjectStatus =
  | "DRAFT"
  | "IN_PROGRESS"
  | "AUDIT_PENDING"
  | "VALIDATED"
  | "COMPLETED"
  | "ARCHIVED";

export type EnergyClass = "A" | "B" | "C" | "D" | "E" | "F" | "G";

export type InsulationItem = "WALLS" | "FLOOR" | "ROOF" | "WINDOWS";

export type HeatingStatus = "NEW" | "KEPT" | "REPLACED";

export type ScenarioType = "SCENARIO_1" | "SCENARIO_2" | "SCENARIO_3";

// ============================================================================
// Types de base - Project
// ============================================================================

export type Project = {
  id: string;
  tenant_id: string;
  client_id: string | null;
  name: string;
  status: ProjectStatus;
  module_code: string;
  building_address: string | null;
  total_apartments: number | null;
  data: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  archived_at: string | null;
};

export type ProjectCreate = {
  name: string;
  client_id?: string | null;
  building_address?: string | null;
  total_apartments?: number | null;
  module_code?: string;
  data?: Record<string, unknown>;
};

export type ProjectUpdate = {
  name?: string | null;
  client_id?: string | null;
  building_address?: string | null;
  total_apartments?: number | null;
  status?: ProjectStatus | null;
  data?: Record<string, unknown> | null;
};

export type ProjectResponse = Project;

export type PaginatedProjectsResponse = {
  items: ProjectResponse[];
  total: number;
  page: number;
  page_size: number;
};

// ============================================================================
// Types pour les opérations sur les projets
// ============================================================================

export type BulkDraftsCreate = {
  quantity: number;
  apartment_type: string;
  common_data?: Record<string, unknown>;
};

export type BulkDraftsResponse = {
  created_count: number;
  project_id: string;
  draft_ids: string[];
};

export type PropagateAuditRequest = {
  source_draft_id: string;
  target_draft_ids: string[];
  fields_to_propagate?: string[] | null;
};

export type PropagateAuditResponse = {
  updated_count: number;
  updated_draft_ids: string[];
  skipped_draft_ids: string[];
};

// ============================================================================
// Types BAR-TH-175 - Données d'audit
// ============================================================================

export type InsulationData = {
  item: InsulationItem;
  total_surface: number;
  isolated_surface: number;
  r_value?: number | null;
  isolation_type?: string | null;
  thickness_cm?: number | null;
  coverage_ratio?: number; // Propriété calculée côté backend
};

export type HeatingData = {
  status: HeatingStatus;
  emission_gco2_kwh: number;
  heating_type?: string | null;
  brand?: string | null;
  model?: string | null;
  power_kw?: number | null;
  cop?: number | null;
  scop?: number | null;
};

export type OccupantData = {
  first_name?: string | null;
  last_name?: string | null;
  email?: string | null;
  phone?: string | null;
  is_tenant?: boolean;
  has_provided_consent?: boolean;
  move_in_date?: string | null;
};

export type BarTh175AuditData = {
  // === État initial ===
  initial_energy_class: EnergyClass;
  initial_ghg: number;
  initial_consumption_kwh_m2?: number | null;

  // === État projeté après travaux ===
  projected_energy_class: EnergyClass;
  projected_ghg: number;
  projected_consumption_kwh_m2?: number | null;

  // === Isolation ===
  insulation_items?: InsulationData[];

  // === Chauffage ===
  heating?: HeatingData | null;

  // === Eau chaude sanitaire ===
  hot_water_type?: string | null;
  hot_water_emission_gco2_kwh?: number | null;

  // === Ventilation ===
  ventilation_type?: string | null;

  // === Métadonnées logement ===
  living_area: number;
  apartment_type?: string | null;
  apartment_number?: number | null;
  floor_level?: number | null;
  construction_year?: number | null;
  nb_rooms?: number | null;

  // === Scénario de travaux ===
  scenario_number?: number | null;
  scenario_type?: ScenarioType | null;
  estimated_cost?: number | null;
  estimated_savings_per_year?: number | null;

  // === Occupant (optionnel pour bailleurs) ===
  occupant?: OccupantData | null;

  // === Champs pour OCR (préparation future) ===
  audit_document_url?: string | null;
  audit_ocr_data?: Record<string, unknown> | null;
  audit_date?: string | null;
  auditor_name?: string | null;
  auditor_certification?: string | null;
};

// ============================================================================
// Types de validation BAR-TH-175
// ============================================================================

export type BarTh175ValidationResult = {
  is_eligible: boolean;
  errors: string[];
  warnings: string[];
  class_jump: number;
  class_jump_valid: boolean;
  insulation_count: number;
  insulation_valid: boolean;
  ghg_reduction_valid: boolean;
  heating_valid: boolean;
};

// ============================================================================
// Types pour les statistiques
// ============================================================================

export type ProjectStats = {
  project_id: string;
  project_name: string;
  status: ProjectStatus;
  total_apartments: number | null;
  drafts_count: number;
  client_id: string | null;
  building_address: string | null;
};
